### PR TITLE
twister: handler: optimize handler testsuite check

### PR DESCRIPTION
**Pull request by hakehuang**
in many platforms, before  the debuger flash, the former testsuite many already executed, so we will see some old testsuite name. and we should not compare them directly, instead if we compare in reverse order then we can avoid such mis-judge.

see below log

```
DEBUG   - DEVICE: ------ TESTSUITE SUMMARY END ------
DEBUG   - DEVICE:
DEBUG   - DEVICE: ===================================================================
DEBUG   - DEVICE: RunID: c16329269d819466809d81610753a0fa
DEBUG   - DEVICE: PROJECT EXECUTION SUCCESSFUL
DEBUG   - -- west flash: using runner linkserver
-- runners.linkserver: LinkServer: /usr/local/LinkServer/LinkServer, version v1.4.85

DEBUG   - Expected suite names:['lib_heap_align', 'lib_heap_align']
DEBUG   - Detected suite names:['onoff_api', 'lib_heap_align', 'lib_heap_align']
DEBUG   - Test suite names were not printed or some of them in output do not correspond with expected: ['lib_heap_align', 'lib_heap_align']
DEBUG   - run status: frdm_mcxn947/mcxn947/cpu0/tests/lib/heap_align/libraries.heap_align failed
INFO    - 1207/1745 frdm_mcxn947/mcxn947/cpu0 tests/lib/heap_align/libraries.heap_align           FAILED Testsuite mismatch (device: #1, 7.110s)
ERROR   - see: /home/ubuntu/nxp/frdm_mcxn947/temp/frdm_mcxn947_mcxn947_cpu0_twister_test_only/twister-out/frdm_mcxn947_mcxn947_cpu0/tests/lib/heap_a


nners.linkserver: LinkServer: /usr/local/LinkServer/LinkServer, version v1.4.85

DEBUG   - Expected suite names:['hash_map', 'hash_map']
DEBUG   - Detected suite names:['hash*** Booting Zephyr OS build v3.6.0-2211-g25812289779f ***', 'hash_map', 'hash_map', 'hash_map', 'hash_map', 'hash_map', 'hash_map', 'hash_map', 'hash_map', 'hash_map', 'hash_map', 'hash_map', 'hash_map', 'hash_map', 'hash_map', 'hash_map']
DEBUG   - Test suite names were not printed or some of them in output do not correspond with expected: ['hash_map', 'hash_map']
DEBUG   - run status: frdm_mcxn947/mcxn947/cpu0/tests/lib/hash_map/libraries.hash_map.separate_chaining.djb2 failed
INFO    - 1226/1745 frdm_mcxn947/mcxn947/cpu0 tests/lib/hash_map/libraries.hash_map.separate_chaining.djb2  FAILED Testsuite mismatch (device: #1, 6.896s)
ERROR   - see: /home/ubuntu/nxp/frdm_mcxn947/temp/frdm_mcxn947_mcxn947_cpu0_twister_test_only/twister-out/frdm_mcxn947_mcxn947_cpu0/tests/lib/hash_map/libr

```

### CODE REVIEW

File: scripts/pylib/twister/twisterlib/handlers.py [hunk 1]
```diff
@@ -133,10 +133,12 @@ def _verify_ztest_suite_name(self, harness_state, detected_suite_names, handler_
             return
         if not detected_suite_names:
             self._missing_suite_name(expected_suite_names, handler_time)
-        for detected_suite_name in detected_suite_names:
-            if detected_suite_name not in expected_suite_names:
+            return
+        # compare the expect and detect from end one by one without order
+        _d_suite = detected_suite_names[-len(expected_suite_names):]
+        if set(_d_suite) != set(expected_suite_names):
+            if not set(_d_suite).issubset(set(expected_suite_names)):
```
Comment (LukaszMrugala) [inline scripts/pylib/twister/twisterlib/handlers.py:140]: Are the erroneous suite names always in the front, or is it just highly likely?

Additionally, in the example you've provided, suite names repeat. This information will be lost when using `set()`. Will this have any negative repercussions? Or should we use `all([suite_name in expected_suite_names for suite_name in _d_suite])`- instead?

Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/handlers.py:140]: > Are the erroneous suite names always in the front, or is it just highly likely?
yes. as those are coming from last case.





> This information will be lost when using `set()`. Will this have any negative repercussions? Or should we use `all([suite_name in expected_suite_names for suite_name in _d_suite])`- instead?

with the first commit in this PR, the repeat suite name is removed if they are intrduced by repeated scan. the thing code comes to this condition is there are cases in cmsis_dsp test, which you expected multiply test suites but only one detected, due to build time flag, one one type of cases are executed.

Comment (golowanow) [reply scripts/pylib/twister/twisterlib/handlers.py:140]: when the detected names have more items than expected and the last items are the same, doesn't the new code lead to a false positive for the expected test ?

Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/handlers.py:140]: @golowanow , the has been filter out by  `if set(_d_suite) != set(expected_suite_names):`

**Comment by LukaszMrugala** [comment]
While this looks like it's working, it seems like a band-aid solution. Can't we fix the underlying problem of reading old data instead?

**Comment by hakehuang** [comment]
> While this looks like it's working, it seems like a band-aid solution. Can't we fix the underlying problem of reading old data instead?

So far I did not find a good soultion, it is a debugger + soc issue, usually the debugger will connect to the SOC and halts the core, however in many our low end soc, this process is slow.

**Comment by hakehuang** [comment]
@golowanow  please check. Thanks

**Comment by hakehuang** [comment]
@gbarkadiusz please help to review, thanks.

**Review by gbarkadiusz — APPROVED** [review]
It seems ok, assuming that the first elements should always be dropped, and removing the duplicates by set() is proper as well.

PS> I'd like to ask about the old data. Where the old data are collecting... As @LukaszMrugala said. It will be better solution to flush old data.

**Comment by hakehuang** [comment]
> PS> I'd like to ask about the old data. Where the old data are collecting... As @LukaszMrugala said. It will be better solution to flush old data.

this is a platform issue, the old data is from former case, the sequence is debug connect the board -> reset core -> halt, this is window in the reset -> halt, the core get chance to run some old code, which outputs the wrong test suitename header. if we want to fix it, we can do,
1. erase every time before test, which cause a lot of time in CI.
2. fix the reset core and halt speed, which is a proble with debug tools, and need work with board by board, the efforts is there.
given above, I would recommand to use current pr solution.

**Comment by gbarkadiusz** [comment]
> > PS> I'd like to ask about the old data. Where the old data are collecting... As @LukaszMrugala said. It will be better solution to flush old data.
> 
> this is a platform issue, the old data is from former case, the sequence is debug connect the board -> reset core -> halt, this is window in the reset -> halt, the core get chance to run some old code, which outputs the wrong test suitename header. if we want to fix it, we can do,
> 
> 1. erase every time before test, which cause a lot of time in CI.
> 2. fix the reset core and halt speed, which is a proble with debug tools, and need work with board by board, the efforts is there.
>    given above, I would recommand to use current pr solution.

Thank you for additional information. So, this pr solution seems ok.

**Review by LukaszMrugala — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**